lockfile/src/parsers/gem.rs

```diff
@@ -1,24 +1,196 @@
-use super::*;
+use nom::branch::alt;
+use nom::bytes::complete::{tag, take_until};
+use nom::character::complete::{line_ending, none_of, space0};
+use nom::combinator::recognize;
+use nom::error::{VerboseError, VerboseErrorKind};
+use nom::multi::{many1, many_till};
+use nom::sequence::{delimited, tuple};
+use nom::Err as NomErr;
 
-pub fn parse(input: &str) -> Result<&str, Vec<PackageDescriptor>> {
-    let (input, _) = gem_header(input)?;
-    let (i, consumed) = specs(input)?;
-    let pkgs = consumed.lines().filter_map(package).collect::<Vec<_>>();
-    Ok((i, pkgs))
+use crate::parsers::{take_till_blank_line, Result};
+use crate::{Package, PackageVersion, ThirdPartyVersion};
+
+const DEFAULT_REGISTRY: &str = "https://rubygems.org/";
+
+#[derive(Debug)]
+struct Section<'a> {
+    section_type: SectionType,
+    content: &'a str,
 }
 
-fn gem_header(input: &str) -> Result<&str, &str> {
-    let (input, _) = recognize(take_until("GEM"))(input)?;
-    recognize(tuple((tag("GEM"), line_ending)))(input)
+impl<'a> Section<'a> {
+    /// Parse lockfile into dependency sections.
+    fn from_lockfile(mut input: &'a str) -> Result<&str, Vec<Self>> {
+        let mut sections = Vec::new();
+
+        while !input.is_empty() {
+            // Find next section head.
+            let header_result = recognize(many_till(
+                take_till_line_end,
+                alt((tag("GEM"), tag("GIT"), tag("PATH"))),
+            ))(input);
+
+            // Stop if no more headers can be found.
+            let (new_input, consumed) = match header_result {
+                Ok(header_result) => header_result,
+                Err(_) => break,
+            };
+
+            // Check for type of section head.
+            let section_type = if consumed.ends_with("GEM") {
+                SectionType::Gem
+            } else if consumed.ends_with("GIT") {
+                SectionType::Git
+            } else if consumed.ends_with("PATH") {
+                SectionType::Path
+            } else {
+                break;
```

Comment:
This should be unreachable code, right? If it ever becomes reachable, I don't want it to be silently ignored.

```suggestion
                unreachable!();
```

Returning an error would also be a decent option, but nom errors are weird to make...